src/BatteryStateOfChargeService.cpp

```diff
@@ -35,29 +36,49 @@ namespace
 BatteryStateOfChargeService::BatteryStateOfChargeService(double initialStateOfChargePercent)
 : initialStateOfChargePercent_(initialStateOfChargePercent)
 {
+
 }
 
 BatteryStateOfChargeService::~BatteryStateOfChargeService()
 {
+
 }
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalcurrent;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (newcurrent<=0)
+        qDebug()<<"is charging";
+    else
+        return false;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    return et;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
-{
+{   
+    newcurrent = batteryData.current;
+    newtime = batteryData.time;
+    s1 = QTime(0,0,0).secsTo(newtime);
+    s2 = QTime(0,0,0).secsTo(oldtime);
+    timeused = s1-s2;
+    totalcurrent = newcurrent - oldcurrent;
+    double avg = (fabs(newcurrent)+fabs(oldcurrent)) *3600 / timeused;
+    int tneed =(BATTERY_AMP_HOUR_CAPACITY - newcurrent)*3600/avg;
```

Comment:
You're using `3600` quite often here and makes it a little difficult to tell what's actually going on in those lines. What would be better is making it a constant. We usually keep our constants in the `.cpp` file in an anonymous namespace at the top.

```c++
namespace 
{
    const int SOME_CONVERSION_FACTOR = 3600;
}
```